File: src/differential_equations_undetermined_coefficients_medium.py
"""
differential_equations - undetermined_coefficients (medium)
Generated: 2026-02-22T05:27:42.285819
"""

from problem_utils import *

def generate():
    # Method of undetermined coefficients for second-order linear ODEs
    # Standard difficulty: 1300-1600 range
    
    problem_type = choice(['exponential', 'polynomial', 'trig', 'mixed'])
    
    if problem_type == 'exponential':
        # y'' + ay' + by = Ce^(rx) where r is not a root of characteristic equation
        a_coeff = nonzero(-4, 4)
        b_coeff = nonzero(-6, 6)
        
        # Characteristic equation: m^2 + a*m + b = 0
        char_roots = solve(m**2 + a_coeff*m + b_coeff, m)
        
        # Choose r different from characteristic roots
        r_val = nonzero(-3, 3)
        if char_roots:
            while any(abs(complex(r_val - root)) < 0.1 for root in char_roots):
                r_val = nonzero(-3, 3)
        
        C_val = nonzero(-5, 5)
        
        question = f"Find a particular solution to the differential equation: $$y'' {'+' if a_coeff > 0 else ''}{a_coeff if abs(a_coeff) != 1 else ('-' if a_coeff == -1 else '')}y' {'+' if b_coeff > 0 else ''}{b_coeff}y = {C_val}e^{{{r_val}x}}$$"
        
        # Particular solution: y_p = A*e^(rx)
        A = symbols('A')
        y_p = A * exp(r_val * x)
        y_p_prime = diff(y_p, x)
        y_p_double_prime = diff(y_p_prime, x)
        
        lhs = y_p_double_prime + a_coeff * y_p_prime + b_coeff * y_p
        lhs_simplified = simplify(lhs)
        
        # Solve for A
        A_val = solve(Eq(lhs_simplified, C_val * exp(r_val * x)), A)[0]
        
        ans = A_val * exp(r_val * x)
        
        solution = steps(
            f"Assume particular solution of form $y_p = Ae^{{{r_val}x}}$",
            f"Compute derivatives: $y_p' = {latex(diff(A * exp(r_val * x), x))}$ and $y_p'' = {latex(diff(diff(A * exp(r_val * x), x), x))}$",
            f"Substitute into DE: ${latex(lhs_simplified)} = {C_val}e^{{{r_val}x}}$",
            f"Simplify: ${latex(factor(lhs_simplified / exp(r_val * x)))}A = {C_val}$",
            f"Solve: $A = {latex(A_val)}$",
            f"Particular solution: $y_p = {latex(ans)}$"
        )
        
        difficulty = (1300, 1450)
        
    elif problem_type == 'polynomial':
        # y'' + ay' + by = polynomial (constant term non-zero ensures no repeated root issues)
        a_coeff = nonzero(-3, 3)
        b_coeff = nonzero(-5, 5)
        
        # Make sure b != 0 (otherwise 0 is a root)
        while b_coeff == 0:
            b_coeff = nonzero(-5, 5)
        
        poly_deg = choice([0, 1, 2])
        
        if poly_deg == 0:
            C_val = nonzero(-8, 8)
            rhs = C_val
            
            question = f"Find a particular solution to: $$y'' {'+' if a_coeff > 0 else ''}{a_coeff if abs(a_coeff) != 1 else ('-' if a_coeff == -1 else '')}y' {'+' if b_coeff > 0 else ''}{b_coeff}y = {C_val}$$"
            
            # y_p = A (constant)
            A_val = Rational(C_val, b_coeff)
            ans = A_val
            
            solution = steps(
                f"Assume particular solution $y_p = A$ (constant)",
                f"Then $y_p' = 0$ and $y_p'' = 0$",
                f"Substitute: ${b_coeff}A = {C_val}$",
                f"Solve: $A = {latex(A_val)}$",
                f"Particular solution: $y_p = {latex(ans)}$"
            )
            difficulty = (1300, 1400)
            
        elif poly_deg == 1:
            c1 = nonzero(-5, 5)
            c0 = nonzero(-5, 5)
            rhs = c1*x + c0
            
            question = f"Find a particular solution to: $$y'' {'+' if a_coeff > 0 else ''}{a_coeff if abs(a_coeff) != 1 else ('-' if a_coeff == -1 else '')}y' {'+' if b_coeff > 0 else ''}{b_coeff}y = {latex(rhs)}$$"
            
            # y_p = Ax + B
            A, B = symbols('A B')
            y_p = A*x + B
            y_p_prime = diff(y_p, x)
            y_p_double_prime = diff(y_p_prime, x)
            
            lhs = y_p_double_prime + a_coeff * y_p_prime + b_coeff * y_p
            lhs_expanded = expand(lhs)
            
            # Collect coefficients
            coeff_system = [
                Eq(lhs_expanded.coeff(x, 1), c1),
                Eq(lhs_expanded.coeff(x, 0), c0)
            ]
            
            sol = solve(coeff_system, [A, B])
            ans = sol[A]*x + sol[B]
            
            solution = steps(
                f"Assume $y_p = Ax + B$",
                f"Then $y_p' = A$ and $y_p'' = 0$",
                f"Substitute: ${latex(lhs_expanded)} = {latex(rhs)}$",
                f"Match coefficients: $A = {latex(sol[A])}$, $B = {latex(sol[B])}$",
                f"Particular solution: $y_p = {latex(ans)}$"
            )
            difficulty = (1400, 1550)
            
        else:  # poly_deg == 2
            c2 = nonzero(-3, 3)
            c1 = nonzero(-4, 4)
            c0 = nonzero(-4, 4)
            rhs = c2*x**2 + c1*x + c0
            
            question = f"Find a particular solution to: $$y'' {'+' if a_coeff > 0 else ''}{a_coeff if abs(a_coeff) != 1 else ('-' if a_coeff == -1 else '')}y' {'+' if b_coeff > 0 else ''}{b_coeff}y = {latex(rhs)}$$"
            
            # y_p = Ax^2 + Bx + C
            A, B, C = symbols('A B C')
            y_p = A*x**2 + B*x + C
            y_p_prime = diff(y_p, x)
            y_p_double_prime = diff(y_p_prime, x)
            
            lhs = y_p_double_prime + a_coeff * y_p_prime + b_coeff * y_p
            lhs_expanded = expand(lhs)
            
            coeff_system = [
                Eq(lhs_expanded.coeff(x, 2), c2),
                Eq(lhs_expanded.coeff(x, 1), c1),
                Eq(lhs_expanded.coeff(x, 0), c0)
            ]
            
            sol = solve(coeff_system, [A, B, C])
            ans = sol[A]*x**2 + sol[B]*x + sol[C]
            
            solution = steps(
                f"Assume $y_p = Ax^2 + Bx + C$",
                f"Compute: $y_p' = 2Ax + B$, $y_p'' = 2A$",
                f"Substitute and collect like terms",
                f"Match coefficients to get $A = {latex(sol[A])}$, $B = {latex(sol[B])}$, $C = {latex(sol[C])}$",
                f"Particular solution: $y_p = {latex(ans)}$"
            )
            difficulty = (1450, 1600)
    
    else:  # trig or mixed
        a_coeff = nonzero(-3, 3)
        b_coeff = nonzero(-6, 6)
        
        omega = choice([1, 2])
        C_val = nonzero(-4, 4)
        
        trig_func = choice([sin, cos])
        trig_name = 'sin' if trig_func == sin else 'cos'
        
        question = f"Find a particular solution to: $$y'' {'+' if a_coeff > 0 else ''}{a_coeff if abs(a_coeff) != 1 else ('-' if a_coeff == -1 else '')}y' {'+' if b_coeff > 0 else ''}{b_coeff}y = {C_val}\\{trig_name}({omega}x)$$"
        
        # y_p = A*cos(wx) + B*sin(wx)
        A, B = symbols('A B')
        y_p = A*cos(omega*x) + B*sin(omega*x)
        y_p_prime = diff(y_p, x)
        y_p_double_prime = diff(y_p_prime, x)
        
        lhs = y_p_double_prime + a_coeff * y_p_prime + b_coeff * y_p
        lhs_expanded = expand(lhs)
        
        # Match with C*trig_func(omega*x)
        if trig_func == sin:
            coeff_system = [
                Eq(lhs_expanded.coeff(sin(omega*x)), C_val),
                Eq(lhs_expanded.coeff(cos(omega*x)), 0)
            ]
        else:
            coeff_system = [
                Eq(lhs_expanded.coeff(cos(omega*x)), C_val),
                Eq(lhs_expanded.coeff(sin(omega*x)), 0)
            ]
        
        sol = solve(coeff_system, [A, B])
        ans = sol[A]*cos(omega*x) + sol[B]*sin(omega*x)
        
        solution = steps(
            f"Assume $y_p = A\\cos({omega}x) + B\\sin({omega}x)$",
            f"Compute derivatives and substitute into DE",
            f"Match coefficients of $\\sin({omega}x)$ and $\\cos({omega}x)$",
            f"Solve system: $A = {latex(sol[A])}$, $B = {latex(sol[B])}$",
            f"Particular solution: $y_p = {latex(ans)}$"
        )
        difficulty = (1450, 1600)
    
    return problem(
        question=question,
        answer=ans,
        difficulty=difficulty,
        topic="differential_equations/undetermined_coefficients",
        solution=solution,
        calculator="scientific"
    )

emit(generate())

File: src/differential_equations_undetermined_coefficients_medium.jl
# differential_equations - undetermined_coefficients (medium)
# Generated: 2026-03-08T20:54:48.754264

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x t begin
    set_topic!("differential_equations/undetermined_coefficients")
    
    # Choose problem type
    problem_type = choice([:exponential, :polynomial, :trig, :mixed])
    
    if problem_type == :exponential
        # y'' + ay' + by = Ce^(kx)
        a = nonzero(-6, 6)
        b = nonzero(-8, 8)
        C = nonzero(-12, 12)
        k = nonzero(-5, 5)
        
        # Check if k is a root of characteristic equation r^2 + ar + b = 0
        char_roots = solve(t^2 + a*t + b ~ 0, t)
        is_root = any(r -> r == k, char_roots)
        
        if is_root
            yp_form = "Axe^{$(k)x}"
            yp = "A*x*exp($(k)*x)"
        else
            yp_form = "Ae^{$(k)x}"
            yp = "A*exp($(k)*x)"
        end
        
        rhs = "$(C)e^{$(k)x}"
        
        question = "Find the form of a particular solution for the differential equation \$\$y'' + $(a)y' + $(b)y = $(rhs)\$\$ using the method of undetermined coefficients."
        answer = yp_form
        
        solution_steps = steps(
            "Homogeneous equation: \$y'' + $(a)y' + $(b)y = 0\$",
            "Characteristic equation: \$r^2 + $(a)r + $(b) = 0\$",
            is_root ? "Since \$k = $(k)\$ is a root of the characteristic equation, multiply by \$x\$" : "Since \$k = $(k)\$ is not a root of the characteristic equation, use standard form",
            sol("Particular solution form", yp_form)
        )
        
        difficulty = is_root ? (1400, 1600) : (1200, 1400)
        time_val = is_root ? 180 : 120
        
    elseif problem_type == :polynomial
        # y'' + ay' + by = polynomial
        a = nonzero(-6, 6)
        b = nonzero(-8, 8)
        
        degree = choice([1, 2])
        
        if degree == 1
            C1 = nonzero(-15, 15)
            C2 = nonzero(-15, 15)
            rhs = "$(C1)x + $(C2)"
            
            # Check if 0 is root of characteristic equation
            if b == 0
                yp_form = "Ax^2 + Bx"
                multiplier = "x"
            else
                yp_form = "Ax + B"
                multiplier = "1"
            end
        else
            C1 = nonzero(-10, 10)
            C2 = nonzero(-15, 15)
            C3 = nonzero(-15, 15)
            rhs = "$(C1)x^2 + $(C2)x + $(C3)"
            
            if b == 0
                yp_form = "Ax^3 + Bx^2 + Cx"
                multiplier = "x"
            else
                yp_form = "Ax^2 + Bx + C"
                multiplier = "1"
            end
        end
        
        question = "Find the form of a particular solution for \$\$y'' + $(a)y' + $(b)y = $(rhs)\$\$ using undetermined coefficients."
        answer = yp_form
        
        solution_steps = steps(
            "The right-hand side is a polynomial of degree $(degree)",
            "Characteristic equation: \$r^2 + $(a)r + $(b) = 0\$",
            b == 0 ? "Since \$r = 0\$ is a root, multiply by \$x\$" : "Since \$r = 0\$ is not a root, use standard polynomial form",
            sol("Particular solution form", yp_form)
        )
        
        difficulty = (1300, 1500)
        time_val = 150
        
    elseif problem_type == :trig
        # y'' + ay' + by = C*cos(kx) or C*sin(kx)
        a = nonzero(-6, 6)
        b = randint(1, 12)  # positive to avoid resonance complexity
        C = nonzero(-12, 12)
        k = randint(1, 5)
        
        trig_func = choice(["cos", "sin"])
        rhs = "$(C)\\\$(trig_func)($(k)x)"
        
        # Check if ±ki are roots of characteristic equation
        # r^2 + ar + b = 0, roots are complex if discriminant < 0
        discriminant = a^2 - 4*b
        
        if discriminant < 0
            # Complex roots r = (-a ± sqrt(disc))/2 = -a/2 ± i*sqrt(-disc)/2
            imag_part = sqrt(-discriminant) / 2
            is_resonant = abs(imag_part - k) < 0.01
        else
            is_resonant = false
        end
        
        if is_resonant
            yp_form = "x(A\\cos($(k)x) + B\\sin($(k)x))"
        else
            yp_form = "A\\cos($(k)x) + B\\sin($(k)x)"
        end
        
        question = "Find the form of a particular solution for \$\$y'' + $(a)y' + $(b)y = $(rhs)\$\$ using undetermined coefficients."
        answer = yp_form
        
        solution_steps = steps(
            "The right-hand side is \$$(rhs)\$",
            "For trigonometric forcing, use both sine and cosine terms",
            is_resonant ? "Resonance detected: multiply by \$x\$" : "No resonance with homogeneous solution",
            sol("Particular solution form", yp_form)
        )
        
        difficulty = is_resonant ? (1500, 1700) : (1300, 1500)
        time_val = is_resonant ? 180 : 150
        
    else  # mixed
        # y'' + ay' + by = polynomial + exponential
        a = nonzero(-5, 5)
        b = nonzero(-8, 8)
        
        C1 = nonzero(-10, 10)
        C2 = nonzero(-12, 12)
        k = nonzero(-4, 4)
        
        rhs = "$(C1)x + $(C2)e^{$(k)x}"
        
        # Check polynomial part
        if b == 0
            poly_part = "Ax^2 + Bx"
        else
            poly_part = "Ax + B"
        end
        
        # Check exponential part
        char_roots = solve(t^2 + a*t + b ~ 0, t)
        is_exp_root = any(r -> r == k, char_roots)
        
        if is_exp_root
            exp_part = "Cxe^{$(k)x}"
        else
            exp_part = "Ce^{$(k)x}"
        end
        
        yp_form = poly_part * " + " * exp_part
        
        question = "Find the form of a particular solution for \$\$y'' + $(a)y' + $(b)y = $(rhs)\$\$ using undetermined coefficients."
        answer = yp_form
        
        solution_steps = steps(
            "The right-hand side has two parts: polynomial and exponential",
            "For the polynomial part: " * (b == 0 ? "multiply by \$x\$ since \$r=0\$ is a root" : "use \$Ax + B\$"),
            "For the exponential part: " * (is_exp_root ? "multiply by \$x\$ since \$k=$(k)\$ is a root" : "use \$Ce^{$(k)x}\$"),
            sol("Particular solution form", yp_form)
        )
        
        difficulty = (1500, 1800)
        time_val = 210
    end
    
    problem(
        question=question,
        answer=answer,
        difficulty=difficulty,
        solution=solution_steps,
        time=time_val,
        answer_type="expression"
    )
end